File: Examenes-EDD-1/2020-Septiembre/Enunciado/AVL.hs
{------------------------------------------------------------------------------
 - Student's name: Paco Molina Cuenca
 -
 - Student's group: Doble Grado Matemáticas e Ingeniería Informática.
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin = undefined

remainingCapacity :: Bin -> Capacity
remainingCapacity = undefined

addObject :: Weight -> Bin -> Bin
addObject = undefined

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity = undefined

height :: AVL -> Int
height = undefined

nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight = undefined

node :: Bin -> AVL -> AVL -> AVL
node = undefined

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft = undefined

addNewBin :: Bin -> AVL -> AVL
addNewBin = undefined
 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst = undefined

addAll:: Capacity -> [Weight] -> AVL
addAll = undefined

toList :: AVL -> [Bin]
toList = undefined

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
-}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: Examenes-EDD-6/2020-Septiembre/Enunciado/Java/AVL.java
/**
 * Student's name: Paco Molina Cuenca
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 */

import dataStructures.list.List;
import dataStructures.list.LinkedList;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
    }

    // adds a new object to this bin
    public void addObject(int weight) {
    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        //  SOLO PARA ALUMNOS SIN EVALUACIÓN CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return weights;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
        }
    }
    
    private static int maxRemainingCapacity(Node node) {
    }

    private static int height(Node node) {
    }

    private Node root; // root of AVL tree

    public AVL() {
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
    }

    private void addNewBinRec(Bin bin, Node root) {
    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
    }

    private void addFirstRec(int initialCapacity, int weight, Node root) {
    }

    public void addAll(int initialCapacity, int[] weights) {
    }

    public List<Bin> toList() {
    }

    private List<Bin> toListRec(Node root) {
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
	}
}
